### ISSUE TITLE
Contacts in contact menu needs to be merged

### ISSUE THREAD

**Issue by LukasReschke** [nextcloud/server#4643]
I do have many people multiple times because I do have them as local user account but also as part of the contacts app, we do need some deduplication happening here.

<img width="452" alt="screen shot 2017-05-01 at 23 17 58" src="https://cloud.githubusercontent.com/assets/878997/25595270/735957f0-2ec4-11e7-9e51-6db2fc7f06c2.png">

**Comment by ChristophWurst** [comment]
Makes sense. Still, this might be tricky depending on the strategy. Do we have anything the uniquely identifies a contact entry?

**Comment by jancborchardt** [comment]
Well … I’m still saying that _some_ form of data sharing between the users management and Contacts app would be useful to fix this and get rid of our many notions of users.

This is not something to be tacked on in the Contacts menu, and share menu, and Mail app, and all the other autocomplete functions.

cc @karlitschek @daita @Henni @irgendwie @skjnldsv @icewind1991

**Comment by rullzer** [comment]
Merging contacts is far from easy. And can often be wrong if you are on a large installation. There is currently no easy way to (reliably) determine that userX (a user on the system) is the same userX some people might have in their contacts.

Also (again large installations) sharing all the system contacts to the users addressbook doesn't seem like a solution.

**Comment by bjornnorman** [comment]
Haven't had time to check the code, but isn't there a chance both userX and the contact are registered with the same email address? If so, at least those matches could - with an acceptable degree of confidence - be merged in the list...?

**Comment by nickvergessen** [comment]
Also by cloud-id?

**Comment by MorrisJobke** [comment]
I guess this is out of scope for 12, right?

**Comment by ChristophWurst** [comment]
> I guess this is out of scope for 12, right?

Yes.

**Comment by jancborchardt** [comment]
Set it to 13 then.

The points by @bjornnorman and @nickvergessen to merge based on email address and cloud-id are good. Additionally we now have the phone number, twitter handle and address in the personal settings, so it’s actually more and more which can overlap and be used as factors to merge on. cc @schiessle

**Comment by skjnldsv** [comment]
> Merging contacts is far from easy. And can often be wrong if you are on a large installation. There is currently no easy way to (reliably) determine that userX (a user on the system) is the same userX some people might have in their contacts.

The Jaro Winkler algorithm is the answer! ;)

**Comment by Arvi89** [comment]
Hello,

If merging is not easy, why not do like mobile phone ? An option to merge contacts, but here that would be to merge contact with a cloud user. In the contact profile, add an option where we can search with a cloud user to merge with.

**Comment by j-ed** [comment]
If a de-duplication cannot be realized using an automatic function, it might be worse to think about the creation of a general feature/app for that job. I'm currently using the [Cardbook extension for Thunderbird](https://github.com/CardBook/CardBook) together with Nextcloud which contains a really good function to merge contact information.
In a [first step](https://addons.cdn.mozilla.net/user-media/previews/full/176/176428.png?modified=1504879594) it shows all duplicate contacts listed in a column view and in a [second step](https://addons.cdn.mozilla.net/user-media/previews/full/176/176429.png?modified=1504879594) a user can decide which information should be kept and merged. Maybe one or other hint can be taken from that project. Just my two cents to the discussion.

**Comment by jancborchardt** [comment]
One basic thing would be to compare email addresses. That's something which would solve it for company/organization setups like ourselves where users are basically the same as Contacts.

**Comment by ChristophWurst** [comment]
> One basic thing would be to compare email addresses. That's something which would solve it for company/organization setups like ourselves where users are basically the same as Contacts.

And fail hard for organizations with shared email accounts and contacts without an email address.

**Comment by jancborchardt** [comment]
Let me spell out the rest of the implementation ;)

- if 2 different contacts have the same email, ignore those from combining
- if either contacts or users do not have an email address, also exclude

All I'm saying it's a way to fix it reasonably for some, if maybe a lot of setups. It would for ours. Why shoot that down?

**Comment by ChristophWurst** [comment]
> Why shoot that down?

That was not my intention, I just wanted to point out my concerns as it's more complex to do that what one would expect.

> if 2 different contacts have the same email, ignore those from combining

*Sounds* reasonable, but will be computationally cost a lot of effort: imagine you've got 300 contacts. Then for each and every contact, you have to check whether any of the other contacts shares the email address, resulting in `n*(n-1)=300*299=87900` comparisons. Depending on the implementation this will take some time.

**Comment by jancborchardt** [comment]
Could it be done in the background? Also, it could always focus on the contacts just in view. Like do it for the contacts which are shown first (currently A–Z, soon most recently interacted with), then when you search through contacts and they are not combined yet, do those which are in view. And I guess cache the stuff.

Alright, then it gets complicated anyway. ;)

**Comment by daita** [comment]
I would approach the issue like said @j-ed. On duplicate (display name in our case) give the user the ability to merge contacts

**Comment by MorrisJobke** [comment]
Move to 14

**Comment by NoMoreNicksLeft** [comment]
The simple answer is, within the contacts app, give us a field where we can select the NC user. Those who do not wish to use this field can leave it blank. Those who do, can do the work manually. I haven't looked at the RFC (is it even IETF?) for ical contacts or whatever it is, but I'm fairly certain it allows custom fields.

Not sure if there are security concerns (I'm an admin on mine, so I don't see it from a non-admin's perspective). Are we leaking sensitive info if a regular user can click that field and get a list of users?

**Comment by peacepenguin** [comment]
Lets perhaps address this in a simpler way. A way that would work for at least some use cases, and provide an alternative workaround path for those waiting on the discussed auto-merge concepts. 

My use case is simple and the problem similar to the OP, with a solution presented that should be easy to implement as a new feature.

In an LDAP-user Nextcloud system, all users and groups are searchable and findable via the Nextcloud share dialogue search. However, email addresses in LDAP do not exist in the "contacts" app. So some users place a copy of the LDAP user info into Contacts so they can sync and see the contact via the built in CardDAV server, and via the Contacts app in nextcloud.

For users that do this, just like the OP, this creates confusion in their Share and Search UIs, because two or more versions of the same object now appear.

I think a simple solution that is easy to implement (at least easier than merge logic), is to create a configuration item like this: "Enable Contacts to be searchable via UI" yes/no  this way, only the "current and correct" LDAP version of the User appears in the search.

This is a similar approach to the global setting that disables auto-complete entirely in the search UIs. 

A further setting to enhance this feature, a user configurable toggle: "Enable Shared Contacts to be searchable via UI" yes/no   this way I can see someone else's contact list they've shared with me in my Thunderbird / android via CardDAV, but I don't see these people when sharing files in Nextcloud Files app.

I think this is a simple feature to implement compared with merge logic, as it simply allows further customization to the "disable auto-complete" feature already available, which is an all-or-nothing approach to this problem that removes too much functionality to be used by most. 

The approach I've presented should be able to eliminate the need for a "merge" approach entirely, as the administrator, or the end user, could decide what works best for them based on the chosen user backend and contact paradigm in use.

**Comment by sokolik87** [comment]
Hi, we have ~50 users and I solve it by deleting rows in **oc_cards_properties** table. I left only those added by LDAP.

**Comment by skjnldsv** [comment]
@sokolik87 that means you're removing all the contacts of your users...

**Comment by sokolik87** [comment]
@skjnldsv not all. there were contacts with email john.doe@mynextcloudhostname.tld and contacts with email john.doe@mycompany.email.tld. So I removed contacts with john.doe@mynextcloudhostname.tld and it solved problem. I don't know if this is right solution, but it is workaround.

**Comment by MorrisJobke** [comment]
This one didn't made it to 15 - let's move it to 16.

**Comment by Leopere** [comment]
Compared to Google's contacts next cloud's is virtually unusable because when I tried to import my contacts from google it kept failing after 100+ contacts so I had to re-import multiple times and also magically now I have like 6 copies of some of them but not all of them and there is zero batch management so now I have to manually delete like 700 contacts and figure out which of the 615 of them are not duplicates.

**Comment by Spartachetto** [comment]
@rullzer shouldn't this be moved to 19 milestone? This is an useful functionality, so if not already present, it should be there.
Or I am missing some important detail???

**Comment by ChristophWurst** [comment]
> it should be there

See https://nextcloud.com/developer/ for ways to contribute.

**Comment by MrRinkana** [comment]
I second this being an important issue, there is a large chance that one already has contact information to many users on an instance, no matter if nextcloud is used amongst family, friends or professionally. The personal and system address-book will create duplicates.

I speculated about hiding the system address-book but normal contact entries don't seem to list shared files even if the federated cloud id is added, so I kept it on.

Maybe instead of merging one could "link" system users and contacts, that prevents them from being displayed twice but keeps the system users unmodified? Linking would be triggered manually by for example adding the system addresbooks "federated cloud id" into the existing contact..

Otherwise this issue is very similar to https://github.com/nextcloud/contacts/issues/70

**Comment by MrRinkana** [comment]
This one is also relevant/duplicate: https://github.com/nextcloud/contacts/issues/243

Some of the pain of these duplicated entries will be resolved with https://github.com/nextcloud/server/pull/50282

**Comment by MrRinkana** [comment]
For UX reasons it would maybe be enough to clearly differentiate the Nextcloud account in the list, so users are not confused about which entry to click.

**Comment by MrRinkana** [comment]
> Makes sense. Still, this might be tricky depending on the strategy. Do we have anything the uniquely identifies a contact entry?

Nowadays (normal) contacts can have a "federated cloud id". I would propose that **if that federated cloud id points to the instance itself**, the contact will overlay the system user entry - essentially "merging" or "linking" them.

Now how to implement that without having to compare every users contacts to the system address book frequently I do not know as I'm not familiar with how the contacts are stored.. But maybe shifting the burden of linking from a user account entry having to point to a contact - to instead the contact pointing to an user satisfies the original worry @ChristophWurst ? It also nicely isolates it to the users own contacts list..

I suppose one would preferably also have some logic that prevents multiple contacts pointing to the same user account, but I suspect that really is uncommon (an address book with duplicates or that's simply broken) and in that case the person with the weird contact list is probably looking into merging those contacts anyway (which they can now!)..

**Comment by WolfgangFahl** [comment]
The contacts app should not even exists without proper deduplication. This is a bug not a feature request. 
It is a typical problem of not agreeing on a proper persistent identifier for an entity. I still have to debug where my duplicate entries come from but from an information science point of view there a quite a few possible unique identifiers. Others might at least help to get high confidence e.g. if picture, e-mail and phone number are the same there is high likelyhood this is the same contact.

**Comment by mrsakrano** [comment]
New reply to your ticket #11436

Title: Re: [nextcloud/server] Contacts in contact menu needs to be merged (#4643)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by shiribailem** [comment]
Throwing in my suggestion for how this would be implemented in a straightforward fashion.

What to merge isn't a huge issue since those are easy checks, but *how* to merge is definitely a concern. One thing is uniform afaik in the system accounts, and that's setting a cloudid and nextcloud social network url.

What I think should happen is if a contact shares one or both of those with a system contact, then it should be treated as an overlay (show what's in the contact card, but then fill in missing info from the system contact).

It would be a pretty simple check and then merging would be nothing more than copying the id/url over and hiding the system contact.
